### PR TITLE
doc: Adopt intra doc links

### PR DESCRIPTION
**Pull request by jan-auer**
This changes the docs build to nightly, which renders intra doc links. `relay-wstring` already made use of them before, but the had not been rendered on our deployed docs.

#skip-changelog

### CODE REVIEW

**Review by flub — APPROVED** [review]

**Comment by jan-auer** [comment]
I'm going to change all occurrences, and then merge this.

File: relay-common/src/metrics.rs [hunk 1]
```diff
@@ -407,6 +404,8 @@ pub trait GaugeMetric {
 }
 
 /// Emits a metric.
+///
+/// See [module-level documentation](metrics/index.html) for examples.
```
Comment (Swatinem) [inline relay-common/src/metrics.rs]: you can also refer to modules by intra-doc syntax

File: relay-wstring/src/lib.rs [hunk 2]
```diff
@@ -221,11 +221,11 @@ impl WStr {
         index.get_unchecked(self)
     }
 
-    /// Returns a mutable subslice of `Self`.
+    /// Returns a mutable subslice of `WStr`.
     ///
     /// # Safety
     ///
-    /// Lice [Self::get_mut] but this results in undefined behaviour if the subslice is not
+    /// Lice [`WStr::get_mut`] but this results in undefined behavior if the subslice is not
```
Comment (Swatinem) [inline relay-wstring/src/lib.rs]:
```suggestion
    /// Like [`WStr::get_mut`] but this results in undefined behavior if the subslice is not
```

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -13,14 +13,19 @@ jobs:
     name: Cargo Docs
     runs-on: ubuntu-latest
 
+    env:
+      RUSTFLAGS: -Dwarnings
```
Comment (flub) [inline .github/workflows/deploy.yml]: I don't think this has any effect

Comment (jan-auer) [reply .github/workflows/deploy.yml]: We use this successfully [in `sentry-rust`](https://github.com/getsentry/sentry-rust/blob/42ff5d1f32b527d773ff12cd43975b58c602cb44/.github/workflows/ci.yml#L52-L53), but I've never used it with cargo doc before. 

CI does not seem to be starting at the moment, so I ran this locally and you're right: For some reason I don't even see the warning when compiling with `RUSTFLAGS=-Dwarnings RUSTUP_TOOLCHAIN=nightly make doc`.

File: relay-common/src/metrics.rs (hunk 1 above)
Comment (jan-auer) [reply relay-common/src/metrics.rs]: Strange, I tried this in the top-level Relay binary, but it didn't work there. Let me try again.

File: .github/workflows/deploy.yml (hunk 3 above)
Comment (jan-auer) [reply .github/workflows/deploy.yml]: Update it does show the warning (it just came up in my scroll buffer), but it doesn't render as error. This is due to [this open issue](https://github.com/rust-lang/rust/issues/43031).

Comment (jan-auer) [reply .github/workflows/deploy.yml]: Found that it works with `RUSTDOCFLAGS`. Thanks, @flub!

Comment (jan-auer) [reply .github/workflows/deploy.yml]: Also updated to the cargo action as per suggestion in https://github.com/getsentry/relay/pull/804 since that will report the failures.

File: relay-common/src/metrics.rs (hunk 1 above)
Comment (jan-auer) [reply relay-common/src/metrics.rs]: So, I got this to work in cases where the module is imported, but not for extern crates, for instance. Any advice on how to import that?

**Review by flub — APPROVED** [review]

File: .github/workflows/deploy.yml [hunk 4]
```diff
@@ -13,14 +13,19 @@ jobs:
     name: Cargo Docs
     runs-on: ubuntu-latest
 
+    env:
+      RUSTDOCFLAGS: -Dbroken_intra_doc_links
```
Comment (jyn514) [inline .github/workflows/deploy.yml:17]: To avoid killing the build cache, you could use `cargo rustdoc -- -D broken_intra_doc_links`. Since you only ever run `cargo doc` once, it might not matter though.

File: relay/src/main.rs [hunk 5]
```diff
@@ -51,11 +53,13 @@
...
+//! [relay documentation]: https://docs.sentry.io/product/relay/
 //! [`relay-auth`]: ../relay_auth/index.html
 //! [`relay-cabi`]: ../relay_cabi/index.html
 //! [`relay-common`]: ../relay_common/index.html
 //! [`relay-config`]: ../relay_config/index.html
+//! [`relay-ffi`]: ../relay_ffi/index.html
+//! [`relay-ffi-macros`]: ../relay_ffi_macros/index.html
 //! [`relay-filter`]: ../relay_filter/index.html
 //! [`relay-general`]: ../relay_general/index.html
 //! [`relay-quotas`]: ../relay_quotas/index.html
```
Comment (jyn514) [inline relay/src/main.rs:65]: You likely couldn't convert these because `relay` doesn't depend on them directly. See https://github.com/rust-lang/rust/issues/74481.

File: .github/workflows/deploy.yml (hunk 4 above)
Comment (flub) [reply .github/workflows/deploy.yml:17]: this does store the rust cache across runs with the swatinem/rust-cache@v1 action.  How does the env vs the cmdline arg affect the cache?

Comment (jyn514) [reply .github/workflows/deploy.yml:17]: RUSTDOCFLAGS is like RUSTFLAGS: whenever it changes, cargo assumes the whole build cache must be invalidated and effectively runs `rm -rf target/doc`. If you pass flags through `rustdoc`, it will only rebuild your crates, not your dependencies.

That said, I don't know how to make `cargo rustdoc` work with workspaces.

Comment (jan-auer) [reply .github/workflows/deploy.yml:17]: That and it doesn't offer `--no-deps`, at least from what I remember. That's why we went with RUSTDOCFLAGS. It works well, since we store the build cache only between docs builds, so the flags never change between builds and the cache is not pruned.
